### PR TITLE
gh-99240: Reset pointer to NULL when the pointed memory is freed in argument parsing

### PR DESCRIPTION
**Pull request by colorfulappl**
In the family of `_PyArg_Parse` functions, pointers can be passed in as arguments to store the parsed PyArgs.
If a pointer is assigned to newly allocated memory and the memory is freed as a result of an error while parsing the subsequent PyArgs, the pointer should be set to NULL; otherwise, it leaks the freed memory.

* Issue: gh-99240

### CODE REVIEW

**Comment by colorfulappl** [comment]
This fix corresponds to the second bullet

> If function _PyArg_ParseStack parses failed, assign all the parsed arguments to "NULL" after they are freed, this should be done in _PyArg_ParseStack.

in my comment in #99240 .

File: Python/getargs.c [hunk 1]
```diff
@@ -202,8 +202,10 @@ _PyArg_VaParse_SizeT(PyObject *args, const char *format, va_list va)
 static int
 cleanup_ptr(PyObject *self, void *ptr)
 {
-    if (ptr) {
-        PyMem_Free(ptr);
+    void **pptr = (void **)ptr;
+    if (*pptr) {
```
Comment (kumaraditya303) [inline Python/getargs.c]: Checking for NULL is redundant. See https://docs.python.org/3/c-api/memory.html?highlight=pymem_free#c.PyMem_Free

**Comment by kumaraditya303** [comment]
Can you add a test for this?

File: Modules/_testcapi/getargs.c [hunk 2]
```diff
@@ -854,6 +854,23 @@ getargs_s_hash_int2(PyObject *self, PyObject *args, PyObject *kwargs)
...
+    char *b = NULL;
+
+    if (!PyArg_ParseTuple(args, "eses", "idna", &a, "idna", &b)) {
+        if (a || b) {
+            PyErr_Clear();
+            PyErr_SetString(PyExc_AssertionError, "Arguments is not cleared.");
+        }
+        return NULL;
+    }
+    Py_DECREF(a);
```
Comment (kumaraditya303) [inline Modules/_testcapi/getargs.c]: why decref on `char *`? You should call `PyMem_Free` on it.

Comment (colorfulappl) [reply Modules/_testcapi/getargs.c]: Sorry for this naive error, I fix it in the latest commit.

File: Modules/_testcapi/getargs.c [hunk 3]
```diff
@@ -854,6 +854,23 @@ getargs_s_hash_int2(PyObject *self, PyObject *args, PyObject *kwargs)
...
+
+    if (!PyArg_ParseTuple(args, "eses", "idna", &a, "idna", &b)) {
+        if (a || b) {
+            PyErr_Clear();
+            PyErr_SetString(PyExc_AssertionError, "Arguments is not cleared.");
+        }
+        return NULL;
+    }
+    PyMem_FREE(a);
+    PyMem_FREE(b);
```
Comment (kumaraditya303) [inline Modules/_testcapi/getargs.c]:
```suggestion
    PyMem_Free(a);
    PyMem_Free(b);
```

File: Modules/_testcapi/getargs.c [hunk 4]
```diff
@@ -854,6 +854,23 @@ getargs_s_hash_int2(PyObject *self, PyObject *args, PyObject *kwargs)
...
 
+static PyObject *
+gh_99240_clear_args(PyObject *self, PyObject *args) {
+    char *a = NULL;
+    char *b = NULL;
+
+    if (!PyArg_ParseTuple(args, "eses", "idna", &a, "idna", &b)) {
+        if (a || b) {
+            PyErr_Clear();
+            PyErr_SetString(PyExc_AssertionError, "Arguments is not cleared.");
```
Comment (kumaraditya303) [inline Modules/_testcapi/getargs.c]:
```suggestion
            PyErr_SetString(PyExc_AssertionError, "Arguments are not cleared.");
```

**Review by kumaraditya303 — APPROVED** [review]
LGTM

File: Modules/_testcapi/getargs.c [hunk 5]
```diff
@@ -854,6 +854,23 @@ getargs_s_hash_int2(PyObject *self, PyObject *args, PyObject *kwargs)
     Py_RETURN_NONE;
 }
 
+static PyObject *
+gh_99240_clear_args(PyObject *self, PyObject *args) {
```
Comment (erlend-aasland) [inline Modules/_testcapi/getargs.c]:
```suggestion
gh_99240_clear_args(PyObject *self, PyObject *args)
{
```

**Review by erlend-aasland — APPROVED** [review]

**Review by kumaraditya303 — APPROVED** [review]

**Comment by kumaraditya303** [comment]
@colorfulappl Can you create the backport PRs? Thanks

**Pull request merged**